feat: add core models for corporate partner access

Ang-m4 opened the pull request:
## Description

This PR introduces the data core models for managing corporate partner catalogs, their associated courses, and user access. The implementation establishes the foundational database schema and backend structure needed for the corporate access feature, including a flexible wrapper to handle dependencies on the Open edX platform.

### Key Changes

* **Dependency Updates  bdb66e0 :** Updated project requirements to include `django-model-utils` for model enhancements and `pillow` for image field support.

* **CourseOverview Backend Wrapper  6e6d082 :** Created a backend abstraction layer (`edxapp_wrapper`) to dynamically load the `CourseOverview` model. This isolates the plugin from the core platform, making it more robust and easier to test.his work also includes an updated testing configuration. This was necessary to correctly handle dependencies on the core platform and ensure all CI checks complete successfully.

* **Initial Data Modeling 4b8c1c3 :**
    * Added the initial Django models for `CorporatePartner`, `CorporatePartnerCatalog`, `CorporatePartnerCatalogLearner`, and related entities.
    * Modeled the many-to-many relationships between catalogs and courses (`CorporatePartnerCatalogCourse`) and catalogs and users (`CorporatePartnerCatalogLearner`) using the explicit `through` option. This allows for storing additional metadata on the relationship itself, such as `position` and `active` status.

* **Database Migrations 4fdf890 :** Generated the initial database migration to create the necessary tables based on the new models.

### How to Test

To test this feature, you'll need to install the plugin in a running Tutor environment and apply the new database migrations.

**Prerequisites:**

  * A running Tutor instance.
  * The plugin source code mounted as a volume into the LMS container (e.g., at `/app/plugins/`).

**Testing Steps:**

1.  **Access the LMS Shell:**
    Open a shell inside the running LMS container:

    ```bash
    tutor dev exec lms bash
    ```

2.  **Install the Plugin:**
    From within the LMS shell, install the plugin in editable mode. This ensures any code changes are immediately reflected without needing to restart the container.

    ```bash
    pip install -e /app/plugins/openedx-corporate
    ```

    *(Note: Adjust the path `/app/plugins/openedx-corporate` to match your specific volume mount location.)*

3.  **Apply Database Migrations:**
    Run the migrations for the `corporate_partner_access` app to create the new tables in the database.

    ```bash
    ./manage.py lms migrate corporate_partner_access
    ```

4.  **Verify the Migration:**
    After running the command, you should see output confirming that the migrations were applied successfully, as shown below:

  <img width="541" height="101" alt="image" src="https://github.com/user-attachments/assets/2976001f-4f81-4d8d-9daf-afceb8464fce" />

Ang-m4 committed:
feat: update requirements to include django-model-utils and pillow

Ang-m4 committed:
feat: implement course overview backend and update settings

Ang-m4 committed:
feat: add core models for corporate partner access

Ang-m4 committed:
feat: create initial migration for corporate partner models

Alec4r commented on corporate_partner_access/models/partner_access.py:
```diff
@@ -0,0 +1,84 @@
+        on_delete=models.CASCADE,
+        related_name="email_regexes",
+    )
+    regex = models.CharField(
```
2 Comments, 1st: We should add a clean to validate that the regex works well, I mean is a valid regex. 2nd: Change the max_length to 500 just in case we want to use a regex more complex.

Alec4r commented on corporate_partner_access/models/partner_access.py:
```diff
@@ -0,0 +1,84 @@
+        ordering = ["position"]
+
+    def __str__(self):
+        return f"<CorporatePartnerCatalogCourse: {self.course_overview.display_name}>"  # pylint: disable=no-member
```
I'm not sure if we will always get a value from display_name, could you check?, if not, could you add a fallback?

Ang-m4 replied:
I think `display_name` may not always be reliable aswell. It might be better to show the course’s identifier (`id`) instead, as it provides definitive and useful information and is a consistent way to reference any `CourseOverview` instance.

Alec4r commented on corporate_partner_access/models/partner_access.py:
```diff
@@ -0,0 +1,84 @@
+    class Meta:
+        verbose_name = "Corporate Partner Catalog Course"
+        verbose_name_plural = "Corporate Partner Catalog Courses"
+        unique_together = ("catalog", "course_overview")
```
I know this already generates an index, but just to help read the code, we could make it explicit by adding the index value.

Ang-m4 commented on corporate_partner_access/models/catalogs.py:
```diff
@@ -0,0 +1,75 @@
+
+    def get_courses(self):
+        """Returns all CourseOverview objects."""
+        return course_overview().objects.all()
```
I think the best approach here is to avoid using the ABC library and instead raise a `NotImplementedError` in the base method. Django models already support abstract base classes through the abstract = True meta attribute, which conflicts with Python’s ABCMeta.

I’ll update the class to raise the appropriate exception.

Ang-m4 committed:
refactor: decouple the flex_catalog core models and update its definition

- add the flex_catalog app for testing and quality checks
- update the setup.py file to include the flex_catalog logic

Ang-m4 committed:
feat: enhance models attributes and validations

- add an email regex validator
- increased the regex field length
- update the str method for the CatalogCourse model

Ang-m4 committed:
feat: update initial migration for corporate partner models

Ang-m4 committed:
feat: add init files for edx wrapper package

Ang-m4 commented on flex_catalog/models.py:
```diff
@@ -0,0 +1,41 @@
+    slug = models.SlugField(unique=True, blank=True, max_length=255)
+    name = models.CharField(max_length=255, help_text="Human friendly name")
+
+    objects = InheritanceManager()
```
@Alec4r, After reviewing this `FlexibleCatalog` implementation, I realized that using `InheritanceManager` in an abstract model doesn't make sense, since abstract models don’t create a database table and are never instantiated.

My concern is that leaving the class as abstract affects the migration and definition of the `CorporatePartnerCatalog` table, and we might end up too far from the initial proposal. To stay close to the original flex‑catalog design, I propose **removing `abstract = True` from the `Meta` class** and treating `FlexibleCatalog` as a concrete model.

This might save us from complex migrations and model changes in the future. It’s possible that we won’t need any other Catalog models, but by making this change, we will remain aligned with the initial proposal.

We should still keep the `NotImplementedError` exception for debugging purposes, as you suggested.
Please let me know what you think! Maybe this isn’t such a big deal, and we could just remove the `objects` attribute instead.

Ang-m4 committed:
feat: update FlexCatalog model and initial migrations

Alec4r reviewed (approved):
LGFM
